fetch/origin/no-cors.window.js

```diff
@@ -73,6 +71,147 @@ promise_test(async function () {
+  };
+}
+
+function fetchReferrerPolicy(referrerPolicy, destination, fetchMode, expectedOrigin) {
+  return async function () {
+    const stash = token(),
+          referrerPolicyPath = "/fetch/origin/resources/referrer-policy.py";
+          redirectPath = "/fetch/origin/resources/redirect-and-stash.py";
+
+    let fetchUrl =
+        (destination=== "same-origin" ? origins.HTTP_ORIGIN
+                                     : origins.HTTP_REMOTE_ORIGIN) +
+        redirectPath + "?stash=" + stash;
+    let json;
+
+    await new Promise(resolve => {
+      const frame = document.createElement("iframe");
+      frame.src = origins.HTTP_ORIGIN + referrerPolicyPath +
+                  "?referrerPolicy=" + referrerPolicy;
+      self.addEventListener("message", e => {
+        if (e.data === "action") {
+          fetch(fetchUrl, { mode: fetchMode, method: "POST" })
+            .then(() => fetch(redirectPath + "?dump&stash=" + stash))
+            .then(rsp => rsp.json())
+            .then(data=>{
+              json = data;
+              frame.remove();
+              resolve();
+            });
+        }
+      }, { once: true });
+      document.body.appendChild(frame);
+    });
+
+    assert_equals(json[0], expectedOrigin);
+  };
+}
+
+function referrerPolicyTestString(referrerPolicy, destination) {
+  return "Origin header and POST " + destination + " with Referrer-Policy " +
+         referrerPolicy;
+}
+
+[
+  {
+    "policy": "no-referrer",
+    "expectedOriginForSameOrigin": "null",
+    "expectedOriginForCrossOrigin": "null"
+  },
+  {
+    "policy": "same-origin",
+    "expectedOriginForSameOrigin": origins.HTTP_ORIGIN,
+    "expectedOriginForCrossOrigin": "null"
+  },
+  {
+    "policy": "origin-when-cross-origin",
+    "expectedOriginForSameOrigin": origins.HTTP_ORIGIN,
+    "expectedOriginForCrossOrigin": origins.HTTP_ORIGIN
+  },
+  {
+    "policy": "no-referrer-when-downgrade",
+    "expectedOriginForSameOrigin": origins.HTTP_ORIGIN,
+    "expectedOriginForCrossOrigin": origins.HTTP_ORIGIN
+  },
+  {
+    "policy": "unsafe-url",
+    "expectedOriginForSameOrigin": origins.HTTP_ORIGIN,
+    "expectedOriginForCrossOrigin": origins.HTTP_ORIGIN
+  },
+].forEach(testObj => {
+  [
+    {
+      "name": "same-origin",
+      "expectedOrigin": testObj.expectedOriginForSameOrigin
+    },
+    {
+      "name": "cross-origin",
+      "expectedOrigin": testObj.expectedOriginForCrossOrigin
+    }
+  ].forEach(destination => {
+    promise_test(navigationReferrerPolicy(testObj.policy,
+                                          destination.name,
+                                          destination.expectedOrigin),
+                 referrerPolicyTestString(testObj.policy,
+                                          destination.name + " navigation"));
+
+    promise_test(fetchReferrerPolicy(testObj.policy,
+                                     destination.name,
+                                     "cors",
```

Comment:
When we're dealing with CORS (mode is "cors") we shouldn't take the referrer policy into account. That only applies to "no-cors" / "navigate", and "same-origin" I suppose. It's an intentional part of using that protocol that makes it okay for us to leak the origin.

Reply:
Hence, spec for "Serializing a request origin" is going to be one more between 1. and 2. :
if CORS flag is unset, request’s origin is same origin with request’s current URL’s origin, and request’s referrer policy is "no-refererer", return "null"

Reply:
Yeah, something like that is probably reasonable, but I mean that the test for "cors" mode does the wrong thing here.

Reply:
@annevk I update the test. Only honor Referrer-Policy when no-cors/navigation and same-origin now. Please check if it's what we want. Thanks.